engine/src/main/java/org/terasology/engine/persistence/typeHandling/extensionTypes/UISkinTypeHandler.java

```diff
@@ -0,0 +1,35 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.engine.persistence.typeHandling.extensionTypes;
+
+import com.google.common.base.Strings;
+import org.terasology.engine.utilities.Assets;
+import org.terasology.nui.skin.UISkin;
+import org.terasology.nui.skin.UISkinAsset;
+import org.terasology.persistence.typeHandling.StringRepresentationTypeHandler;
+
+import java.util.Optional;
+
+public class UISkinTypeHandler extends StringRepresentationTypeHandler<UISkin> {
+    @Override
+    public String getAsString(UISkin item) {
+        if (item == null) {
+            return "";
+        }
+        // TODO
+        return item.toString();
```

Comment:
@BenjaminAmos "TODO"? Is this not quite right?

Reply:
When this was handled by `AssetTypeHandler`, then this method would return the asset urn for the skin.
https://github.com/MovingBlocks/Terasology/blob/cbc07339ea0fdbfaf1cfe31beee8b0b44a25834e/engine/src/main/java/org/terasology/engine/persistence/typeHandling/extensionTypes/AssetTypeHandler.java#L22-L28
When given a `UISkin` though, there is no direct mapping to the resource that generated it. I wasn't sure how to return the source asset urn from there. Any help with that bit would be appreciated.

It's TODO because I have no idea what the consequences of leaving this unimplemented will be but given that it relates to serialisation, it could cause some issues.

Reply:
@DarkWeird @eviltak any ideas?

Reply:
> @DarkWeird @eviltak any ideas?

1. store reverse link somewhere. 
2. don't save UISkin separately
3. creates common (UISkin/UISkinAsset)TypeHandlerFactory

Reply:
I've given this another attempt, however it's still a work-around. It involves sub-classing `UISkin` to add a `ResourceUrn` to it and then copying the retrieved skin into an instance of that subclass. It's at https://github.com/BenjaminAmos/Terasology/commit/c8ffd14a9325cd5515156aa1fc07f0f03b2bc128. If I can't think of anything better soon, then I'll send a PR for it. Suprisingly though, I couldn't actually find anywhere where this particular method is called. Components don't tend to store `UISkin` elements directly and so-far I have only found deserialisation usage for this type handler. That may be why having this return invalid values doesn't seem to affect gameplay.

Reply:
I can't see how a reverse link would work, as two assets could hypothetically contain the same content yet be from different sources. `AssetTypeHandler` handles all asset types, not just `UISkin`. I haven't thought of any other solutions, so, whilst it's not ideal, I've submiited #4604 to potentially resolve this.